runtime/common/src/main/kotlin/kotlinx/serialization/modules/SerialModuleImpl.kt

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2017-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+@file:Suppress("UNCHECKED_CAST")
+
+package kotlinx.serialization.modules
+
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.isInstanceOf
+import kotlin.collections.set
+import kotlin.reflect.KClass
+
+private typealias SerializersMap = MutableMap<KClass<*>, KSerializer<*>>
+
+/**
+ * A default implementation of [SerialModule]
+ * which uses hash maps to store serializers associated with KClasses.
+ */
+internal class SerialModuleImpl : SerialModule {
+
+    private val classMap: SerializersMap = hashMapOf()
+
+    private val polyMap: MutableMap<KClass<*>, SerializersMap> = hashMapOf()
+    private val inverseClassNameMap: MutableMap<KClass<*>, MutableMap<String, KSerializer<*>>> = hashMapOf()
+
+    internal fun <T : Any> registerSerializer(
+        forClass: KClass<T>,
+        serializer: KSerializer<T>,
+        allowOverwrite: Boolean = false
+    ) {
+        if (!allowOverwrite && forClass in classMap) throw SerializerAlreadyRegisteredException(forClass)
+        classMap[forClass] = serializer
+    }
+
+    internal fun <Base : Any, Sub : Base> registerPolymorphicSerializer(
+        baseClass: KClass<Base>,
+        concreteClass: KClass<Sub>,
+        concreteSerializer: KSerializer<Sub>,
+        allowOverwrite: Boolean = false
+    ) {
+        val name = concreteSerializer.descriptor.name
+        polyMap.getOrPut(baseClass, ::hashMapOf).let { baseClassMap ->
+            if (!allowOverwrite && concreteClass in baseClassMap) throw SerializerAlreadyRegisteredException(
```

Comment:
This is a serious usability issue: we should not throw `SerializerAlreadyRegisteredException` if **the same** serializer is already present.
This will be especially frustrating when using `Module1 + Module2` which have overlapping serializers (e.g. they both include `JavaSerializerModules`)

Reply:
We don't have well-defined equality on serializers (and I think it is not even possible to define it), so our best guess will be referential equality – which may fail if serializers are generic and therefore are classes, not objects.

Reply:
Referential equality is enough here as the best effort.
For the generic implementation, it is a good idea to provide an `equals` as well (we can discuss/fix it separately, not in the scope of this change)

Reply:
I'm not sure this is a good idea since we are introducing a bug/undefined behavior instead of a (little) inconvenience which is obvious to overcome. Any module exported from 3rd party library should not include anything except classes from that library, and this problem will likely never happen.

Reply:
#411